tests/zfs-tests/tests/functional/cli_root/zpool_events/zpool_events_clear.ksh

```diff
@@ -0,0 +1,62 @@
+#!/bin/ksh -p
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2017, loli10K <ezomori.nozomu@gmail.com>. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# 'zpool events -c' should successfully clear events.
+#
+# STRATEGY:
+# 1. Clear all ZFS events
+# 2. Generate some new ZFS events
+# 3. Verify 'zpool events -c' successfully clear new events
+#
+
+verify_runnable "both"
+
+log_assert "'zpool events -c' should successfully clear events."
+
+EVENTS_NUM="$(random 15)"
+
+# 1. Clear all ZFS events
+# This is needed because we may already over the max number or events queued
+# (zfs_zevent_len_max) generated by previous tests: generating $EVENTS_NUM new
+# events and then counting them is racy and leads to failues, so start from 0.
+log_must zpool events -c
+
+# 2. Generate some new ZFS events
+for i in `seq 1 $EVENTS_NUM`; do
+	log_must zpool clear $TESTPOOL
+done
+# Again, this is racy: we may be counting some events less than expected, so
+# we wait a bit to allow the kernel module to process every new event.
+sleep 3
```

Comment:
I actually already implemented a new function `wait_events` in `zpool_events.kshlib` (code not pused yet), but `zpool_events_settle` is objectively better.

Reply:
There seems to be a race condition here, though `zpool_events_poolname.ksh` is the script that is now triggering it: http://build.zfsonlinux.org/builders/Ubuntu%2014.04%20i686%20%28TEST%29/builds/4383/steps/shell_8/logs/log

```
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_events/zpool_events_poolname (run as root) [00:04] [FAIL]
19:05:02.26 ASSERTION: 'zpool events poolname' should only display events from poolname.
19:05:02.27 SUCCESS: truncate -s 268435456 /mnt/newpool.dat
19:05:02.39 SUCCESS: zpool create newpool /mnt/newpool.dat
19:05:02.40 cleared 28 events
19:05:02.41 SUCCESS: zpool events -c
19:05:02.45 SUCCESS: zpool clear testpool
19:05:02.48 SUCCESS: zpool clear testpool
19:05:02.52 SUCCESS: zpool clear testpool
19:05:02.55 SUCCESS: zpool clear testpool
19:05:02.58 SUCCESS: zpool clear testpool
19:05:02.61 SUCCESS: zpool clear testpool
19:05:02.64 SUCCESS: zpool clear testpool
19:05:02.68 SUCCESS: zpool clear testpool
19:05:02.76 SUCCESS: zpool clear testpool
19:05:02.79 SUCCESS: zpool clear testpool
19:05:02.85 SUCCESS: zpool clear testpool
19:05:02.88 SUCCESS: zpool clear testpool
19:05:02.90 SUCCESS: zpool clear testpool
19:05:02.94 SUCCESS: zpool clear testpool
19:05:03.06 SUCCESS: zpool clear testpool
19:05:03.08 SUCCESS: zpool clear newpool
19:05:03.11 SUCCESS: zpool clear newpool
19:05:03.14 SUCCESS: zpool clear newpool
19:05:03.17 SUCCESS: zpool clear newpool
19:05:03.19 SUCCESS: zpool clear newpool
19:05:06.23 NOTE: waited 3 seconds
19:05:06.24 Oct 19 2017 19:05:02.401430000	sysevent.fs.zfs.history_event
19:05:06.24 Oct 19 2017 19:05:02.445430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.477430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.509430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.541430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.573430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.605430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.633430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.669430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.753430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.781430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.837430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.869430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.897430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:02.937430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:03.057430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:03.077430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:03.105430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:03.133430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:03.161430000	sysevent.fs.zfs.config_sync
19:05:06.24 Oct 19 2017 19:05:03.185430000	sysevent.fs.zfs.config_sync
19:05:06.27 Unexpected events (newpool): 6 != 5
19:05:06.27 NOTE: Performing test-fail callback (/usr/share/zfs/zfs-tests/callbacks/zfs_dbgmsg.ksh)
```

Though `zpool clear` is indeed synchronous the fact we can see a "sysevent.fs.zfs.history_event" (which is from the `zpool create newpool /mnt/newpool.dat`) *after* `zpool events -c` seems to confirm event delivery to be asynchronous.

Reply:
Ok, thanks for looking into that. This is going to pose a problem to your count based tests, isn't it? You could get around that for `zpool_events_poolname.ksh` by simply not caring about the number of events. Just ensure that `zpool events pool` output is from `pool`. Maybe something like the following will work in your case?

`zpool events -v newpool | grep "pool ="  | grep -v "pool = \"newpool\""`

Reply:
This race could theoretically affect other tests where we assume 0 events after `zpool events -c` (zpool_events_follow, zpool_events_clear), though is much more difficult to trigger because the other events must be from previous test cases.

I'm going to try your proposed solution locally and force-push when the test seems reliable.